scipy/spatial/_hausdorff.py

```diff
@@ -0,0 +1,53 @@
+#
+# Copyright (C)  Tyler Reddy, Richard Gowers, and Max Linke, 2016
+#
+# Distributed under the same BSD license as Scipy.
+#
+# adapted from scipy's cython version
+import numpy as np
+import random
+#pythran export directed_hausdorff(float64[:,:], float64[:,:], int)
+#pythran export directed_hausdorff(float64[:,:], float64[:,:])
+
+def directed_hausdorff(ar1, ar2, seed=0):
+
+    N1, data_dims = ar1.shape
+    N2 = ar2.shape[0]
+    i_store = j_store = i_ret = j_ret = 0
+
+    # shuffling the points in each array generally increases the likelihood of
+    # an advantageous break in the inner search loop and never decreases the
+    # performance of the algorithm
+    random.seed(seed)
+    resort1 = np.arange(N1)
+    resort2 = np.arange(N2)
+    random.shuffle(resort1)
+    random.shuffle(resort2)
+    ar1 = np.asarray(ar1)[resort1]
+    ar2 = np.asarray(ar2)[resort2]
+
+    cmax = 0
+    for i in range(N1):
+        cmin = np.inf
+        for j in range(N2):
+            d = np.sum((ar1[i] - ar2[j]) ** 2)
+            # faster performance with square of distance
+            # avoid sqrt until very end
+            if d < cmax:  # break out of `for j` loop
+                break
+
+            if d < cmin:  # always true on first iteration of for-j loop
+                cmin = d
+                i_store = i
+                j_store = j
+        else:
+
+            # always true on first iteration of for-j loop, after that only
+            # if d >= cmax
+            if cmin != np.inf and cmin > cmax:
+                cmax = cmin
+                i_ret = i_store
+                j_ret = j_store
+
+    return np.sqrt(cmax), resort1[i_ret], resort2[j_ret]
```

Comment:
It looks like some low-level C calls have been replaced by NumPy operations for Pythran suitability--if this happens eventually & we do actually keep separate source files for Pythran and plain Cython, are we also going to flush the tests through both compiled files, and will this be straightforward to do?

Reply:
I don't think we should aim at maintaining two code base. Both for maintainablity or, as you point out, testability :-)

Reply:
I agree, any duplicate implementation should be temporary or not be done. We see that now with, e.g., `KDTree` and `cKDTree` that have gotten out of sync but we don't get around to cleaning up.